File: aifupan_project/ReviewAnalysis/doc/搜达人/get_sessionid.py


import requests
import time
import json
import random

class Spider(object):
    def __init__(self):
        self.account = 'dengdingnan@jiuyujiaoyu.com'  # 邮箱
        self.password = 'Aa12345678'  # 密码

    def encrypt(self, e):
        if e is None:
            return ""
        # 将字符串转换为UTF-8字节列表（对应JavaScript中的字符编码处理）
        t = []
        for char in e:
            code = ord(char)
            if 0 <= code <= 127:
                t.append(code)
            elif 128 <= code <= 2047:
                t.append(192 | ((code >> 6) & 31))
                t.append(128 | (code & 63))
            else:
                if (2048 <= code <= 55295) or (57344 <= code <= 65535):
                    t.append(224 | ((code >> 12) & 15))
                    t.append(128 | ((code >> 6) & 63))
                    t.append(128 | (code & 63))

        # 对每个字节执行异或操作并转换为十六进制字符串
        n = []
        for byte in t:
            byte &= 255  # 确保字节范围在0-255
            n.append("{0:02x}".format(5 ^ byte))
        return ''.join(n)

    # 生成fp
    def get_fp(self):
        chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
        e = list(chars)
        t = len(e)
        n = str(int(time.time()))  # 当前时间戳转为字符串

        # 使用 Date.now().toString(36) 的模拟，将时间戳转为 base36
        def base36encode(number):
            return format(number, 'x')  # 这里简化处理，Python 中使用 hex 类似于 JS 的 toString(36)

        n = base36encode(int(time.time()))

        r = [''] * 36
        r[8] = r[13] = r[18] = r[23] = "_"
        r[14] = "4"

        for o in range(36):
            if not r[o]:
                i = int(random.random() * t)
                if o == 19:
                    r[o] = e[(3 & i) | 8]
                else:
                    r[o] = e[i]
        fp = "verify_" + n + "_" + "".join(r)
        # print(f"生成的fp: {fp}")
        return fp

    # 登录
    def account_login(self):
        headers = {
            'accept': 'application/json, text/plain, */*',
            'accept-language': 'zh-CN,zh;q=0.9',
            'cache-control': 'no-cache',
            'content-type': 'application/x-www-form-urlencoded',
            'origin': 'https://www.lifexue.com',
            'pragma': 'no-cache',
            'priority': 'u=1, i',
            'referer': 'https://www.lifexue.com/',
            'sec-ch-ua': '"Not)A;Brand";v="8", "Chromium";v="138", "Google Chrome";v="138"',
            'sec-ch-ua-mobile': '?0',
            'sec-ch-ua-platform': '"Windows"',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'cross-site',
            'sec-fetch-storage-access': 'none',
            'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36',
            'x-requested-with': 'XMLHttpRequest',
            'x-tt-passport-csrf-token': '',
        }

        # params = {
        #     'msToken': '',
        #     'X-Bogus': 'DFSzswVLQDGJ3SXFCtLY4rYhjsEI',
        #     '_signature': '_02B4Z6wo00001LvAw9gAAIDDJaRjSkc0gJC7wMdAAEaipojb1HFJXhAdoeY3d9msEYvVHZPhvyX0soU5SGqBUKvuW0kLpG5t2b-9fseWGJFSWjHO.mrmFt2SvxgVHo-I-sjsxYk2JmrCTKO-4c',
        # }

        data = {
            'fp': self.get_fp(),
            'aid': '405096',
            'language': 'zh',
            'account_sdk_source': 'web',
            'mix_mode': '1',
            'service': 'https://www.lifexue.com/',
            'account': self.encrypt(self.account),
            'password': self.encrypt(self.password),
            'captcha_key': '',
        }
        # data = json.dumps(data, separators=(',', ':'))

        response = requests.post('https://sso.oceanengine.com/account_login/v2/', headers=headers, data=data)
        json_data = response.json()
        if json_data['error_code'] == 0:
            user_id = json_data.get('user_id')
            print(f"学号user_id: {user_id}")
            redirect_url = json_data.get('redirect_url')
            print(f"回调接口：{redirect_url}")
            self.login_callback(redirect_url)

        else:
            print(response.text)

    # 登录回调返回cookie
    def login_callback(self, redirect_url):
        headers = {
            'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
            'accept-language': 'zh-CN,zh;q=0.9',
            'cache-control': 'no-cache',
            'pragma': 'no-cache',
            'priority': 'u=0, i',
            'referer': 'https://www.lifexue.com/',
            'sec-ch-ua': '"Not)A;Brand";v="8", "Chromium";v="138", "Google Chrome";v="138"',
            'sec-ch-ua-mobile': '?0',
            'sec-ch-ua-platform': '"Windows"',
            'sec-fetch-dest': 'document',
            'sec-fetch-mode': 'navigate',
            'sec-fetch-site': 'same-origin',
            'sec-fetch-user': '?1',
            'upgrade-insecure-requests': '1',
            'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36',
            # 'cookie': 'x-web-secsdk-uid=a68bc8e5-5c7f-4a1d-bcbc-03e63c69d965; s_v_web_id=verify_mcvetxvk_7BtqVCX7_9EnG_4D25_9okr_rVcw8EWnmYbW; csrf_session_id=00e9b3ae0fe0cc1b8c3aaaa2405175cb; ttcid=c089858b27ab4986912d7e4a8fced5e231; tt_scid=YdEoFh-n-24DSvwWuhq0rZfmOlT3PqUI3AaWklSWSzZID4f9nfl3r2cN01AUDHTb70e6',
        }
        response = requests.get(redirect_url, headers=headers, allow_redirects=False)
        print(response)
        response_cookie = response.cookies.get_dict()
        sessionid = response_cookie.get('sessionid')
        print(f'sessionid: {sessionid}')



if __name__ == '__main__':
    s = Spider()
    s.account_login()

File: aifupan_project/ReviewAnalysis/doc/搜达人/get_sessionid.cs


using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

public class Spider
{
    private readonly string _account = "dengdingnan@jiuyujiaoyu.com";
    private readonly string _password = "Aa12345678";
    private readonly HttpClient _httpClient;
    private readonly HttpClientHandler _handler;

    public Spider()
    {
        // 初始化HttpClientHandler并禁用自动重定向
        _handler = new HttpClientHandler
        {
            AllowAutoRedirect = false,  // 对应Python的allow_redirects=False
            UseCookies = true,
            CookieContainer = new CookieContainer()
        };
        
        _httpClient = new HttpClient(_handler);
        _httpClient.DefaultRequestHeaders.UserAgent.ParseAdd(
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36");
    }

    // 加密方法保持不变
    public string Encrypt(string e)
    {
        if (string.IsNullOrEmpty(e))
            return "";

        var t = new List<byte>();

        foreach (char c in e)
        {
            int code = c;

            if (code >= 0 && code <= 127)
            {
                t.Add((byte)code);
            }
            else if (code >= 128 && code <= 2047)
            {
                t.Add((byte)(192 | ((code >> 6) & 31)));
                t.Add((byte)(128 | (code & 63)));
            }
            else if ((code >= 2048 && code <= 55295) || (code >= 57344 && code <= 65535))
            {
                t.Add((byte)(224 | ((code >> 12) & 15)));
                t.Add((byte)(128 | ((code >> 6) & 63)));
                t.Add((byte)(128 | (code & 63)));
            }
        }

        var n = new List<string>();
        foreach (byte b in t)
        {
            byte xorResult = (byte)(5 ^ b);
            n.Add(xorResult.ToString("x2"));
        }

        return string.Join("", n);
    }

    // GetFp方法保持不变
    public string GetFp()
    {
        const string chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
        var random = new Random();
        long timestamp = (long)(DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)).TotalSeconds;
        
        // 生成base36编码的时间戳
        string n = Base36Encode(timestamp).ToLower();

        var r = new char[36];
        r[8] = '_';
        r[13] = '_';
        r[18] = '_';
        r[23] = '_';
        r[14] = '4';

        for (int o = 0; o < 36; o++)
        {
            if (r[o] == '\0')
            {
                int i = random.Next(chars.Length);
                if (o == 19)
                {
                    r[o] = chars[(3 & i) | 8];
                }
                else
                {
                    r[o] = chars[i];
                }
            }
        }

        return $"verify_{n}_{new string(r)}";
    }

    // Base36Encode方法保持不变
    private string Base36Encode(long number)
    {
        const string chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        if (number == 0) return "0";
        
        var result = new StringBuilder();
        bool isNegative = number < 0;
        
        if (isNegative)
            number = -number;

        while (number > 0)
        {
            result.Append(chars[(int)(number % 36)]);
            number /= 36;
        }

        if (isNegative)
            result.Append('-');

        // 反转结果以获得正确的顺序
        char[] arr = result.ToString().ToCharArray();
        Array.Reverse(arr);
        return new string(arr);
    }

    public async Task AccountLogin()
    {
        // 清除默认请求头，避免冲突
        _httpClient.DefaultRequestHeaders.Clear();
        
        // 设置基本请求头（不包含content-type）
        var headers = new Dictionary<string, string>
        {
            {"accept", "application/json, text/plain, */*"},
            {"accept-language", "zh-CN,zh;q=0.9"},
            {"cache-control", "no-cache"},
            {"origin", "https://www.lifexue.com"},
            {"pragma", "no-cache"},
            {"priority", "u=1, i"},
            {"referer", "https://www.lifexue.com/"},
            {"sec-ch-ua", "\"Not)A;Brand\";v=\"8\", \"Chromium\";v=\"138\", \"Google Chrome\";v=\"138\""},
            {"sec-ch-ua-mobile", "?0"},
            {"sec-ch-ua-platform", "\"Windows\""},
            {"sec-fetch-dest", "empty"},
            {"sec-fetch-mode", "cors"},
            {"sec-fetch-site", "cross-site"},
            {"sec-fetch-storage-access", "none"},
            {"x-requested-with", "XMLHttpRequest"},
            {"x-tt-passport-csrf-token", ""},
            {"user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36"}
        };

        foreach (var header in headers)
        {
            _httpClient.DefaultRequestHeaders.TryAddWithoutValidation(header.Key, header.Value);
        }

        var formData = new FormUrlEncodedContent(new[]
        {
            new KeyValuePair<string, string>("fp", GetFp()),
            new KeyValuePair<string, string>("aid", "405096"),
            new KeyValuePair<string, string>("language", "zh"),
            new KeyValuePair<string, string>("account_sdk_source", "web"),
            new KeyValuePair<string, string>("mix_mode", "1"),
            new KeyValuePair<string, string>("service", "https://www.lifexue.com/"),
            new KeyValuePair<string, string>("account", Encrypt(_account)),
            new KeyValuePair<string, string>("password", Encrypt(_password)),
            new KeyValuePair<string, string>("captcha_key", "")
        });

        // 重要：content-type通过FormUrlEncodedContent自动设置为application/x-www-form-urlencoded
        // 不需要手动添加

        try
        {
            var response = await _httpClient.PostAsync("https://sso.oceanengine.com/account_login/v2/", formData);
            
            // 检查是否有重定向（3xx状态码）
            if (response.StatusCode >= HttpStatusCode.MultipleChoices && 
                response.StatusCode <= HttpStatusCode.TemporaryRedirect)
            {
                Console.WriteLine($"收到重定向响应: {response.StatusCode}");
                if (response.Headers.TryGetValues("Location", out var locations))
                {
                    foreach (var location in locations)
                    {
                        Console.WriteLine($"重定向地址: {location}");
                    }
                }
            }
            
            string responseBody = await response.Content.ReadAsStringAsync();
            Console.WriteLine($"登录响应: {responseBody}");
            
            var jsonData = System.Text.Json.JsonSerializer.Deserialize<System.Text.Json.Nodes.JsonObject>(responseBody);

            if (jsonData != null && jsonData.TryGetPropertyValue("error_code", out var errorCode) && 
                errorCode.GetValue<int>() == 0)
            {
                var userId = jsonData.TryGetPropertyValue("user_id", out var userIdNode) ? userIdNode.ToString() : "未知";
                Console.WriteLine($"学号user_id: {userId}");

                if (jsonData.TryGetPropertyValue("redirect_url", out var redirectUrlNode))
                {
                    string redirectUrl = redirectUrlNode.ToString();
                    Console.WriteLine($"回调接口：{redirectUrl}");
                    await LoginCallback(redirectUrl);
                }
            }
            else
            {
                Console.WriteLine("登录失败: " + responseBody);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"登录请求异常: {ex.Message}");
            Console.WriteLine($"异常详情: {ex.ToString()}");
        }
    }

    public async Task LoginCallback(string redirectUrl)
    {
        // 清除之前的请求头
        _httpClient.DefaultRequestHeaders.Clear();
        
        // 设置回调请求头
        var headers = new Dictionary<string, string>
        {
            {"accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"},
            {"accept-language", "zh-CN,zh;q=0.9"},
            {"cache-control", "no-cache"},
            {"pragma", "no-cache"},
            {"priority", "u=0, i"},
            {"referer", "https://www.lifexue.com/"},
            {"sec-ch-ua", "\"Not)A;Brand\";v=\"8\", \"Chromium\";v=\"138\", \"Google Chrome\";v=\"138\""},
            {"sec-ch-ua-mobile", "?0"},
            {"sec-ch-ua-platform", "\"Windows\""},
            {"sec-fetch-dest", "document"},
            {"sec-fetch-mode", "navigate"},
            {"sec-fetch-site", "same-origin"},
            {"sec-fetch-user", "?1"},
            {"upgrade-insecure-requests", "1"},
            {"user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36"}
        };

        foreach (var header in headers)
        {
            _httpClient.DefaultRequestHeaders.TryAddWithoutValidation(header.Key, header.Value);
        }

        try
        {
            var response = await _httpClient.GetAsync(redirectUrl);
            Console.WriteLine($"回调响应状态码: {response.StatusCode}");

            // 检查是否有重定向
            if (response.StatusCode >= HttpStatusCode.MultipleChoices && 
                response.StatusCode <= HttpStatusCode.TemporaryRedirect)
            {
                Console.WriteLine($"回调请求收到重定向: {response.StatusCode}");
            }

            // 获取响应中的Cookie
            var cookies = _handler.CookieContainer.GetCookies(new Uri(redirectUrl));
            foreach (Cookie cookie in cookies)
            {
                if (cookie.Name == "sessionid")
                {
                    Console.WriteLine($"sessionid: {cookie.Value}");
                    break;
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"回调请求异常: {ex.Message}");
            Console.WriteLine($"异常详情: {ex.ToString()}");
        }
    }

    public static async Task Main(string[] args)
    {
        var spider = new Spider();
        await spider.AccountLogin();
    }
}
